Fixed popups for clustered points (please don't merge until tests are fixed)

OlaKov opened the pull request:
Hi Tom, 

I edited the qgis2web.js file to make pupups work for clustered point layers.  On click/hover the clustered points now show information, instead of the [Objects object]. If 3 points are clustered, the popup will show fields that are not set to 'hidden' for all 3 features (I think this makes the most sense, but it can be changed to "please zoom in"). 

To display how many features are clustered and to get rid of the "undefined" upon high zoom-in, I need to change the _style.js files, which are generated through python olStyleScripts.py, exportStyles function (is this right?). I know exactly what needs to be changed in the .js file, but I'm having a hard time testing the olStyleScripts.py file, when I change it within the plugin folder and try to make a qgis2web preview with QGIS desktop, it doesn't recognize my edits. I will try to figure this out over the next few days and then clustering should work well!

Ola

tomchadwin commented:
Wow. Thanks for this! If I understand you correctly that you don't see any change if you edit Python files, that is because they need to be compiled into .pyc files, which are what is actually run. That compilation happens when the plugin is loaded. You can therefore restart QGIS or use the plugin manager, but there is an easier way. 

Open the plugin manager, *enable experimental plugins*, and find the plugin "Plugin Reloader". Install that, and look for its two icons in the main tool bar. It makes it much quicker and easier to reload the plugin, and hence compile the Python files. 

Once it's installed and configured, use it to reload our plugin as soon as you want to test any change in any Python file. 

Let me know of that works, or if you have any specific questions.

OlaKov commented:
Plugin Reloader works great, thanks!

So for the olStyleScripts, I got rid of: 
if (%(label)s !== null%(labelRes)s) {
        var labelText = String(%(label)s);
    } else {
        var labelText = ""
    }
since it attempts to display feature information on high zoom-in, but it's not handled properly so the "undefined" is displayed. I think it's better to display info through the popup, since it actually checks for hidden fields etc.

I added an 'if statement' that displays the cluster number if the cluster has 2 of more features (I don't think it's necessary to display 1 for a single point").

One issue is that the cluster pixel size is directly connected to the number of clustered feature, so 1000 clustered points will produce a circle that is original size+1000 pixels. I will work on restricting clusters to 20 pixels next, but the python files are more tricky. 

Please feel free to merge once I get the tests fixed, I probably won't get to fixing the "pixel issue" until later in the week, or let me know if you want the code changed.

tomchadwin commented:
OK, thanks for this - I've had a look through. I've used the airports shapefile in the /test folder of the plugin, and enabled clustering.

{object][object] is gone, and I can see the data, which is great. However, only the first record is formatted correctly. The subsequent ones are run together:

![2017-05-15 12_58_13-export to web map](https://cloud.githubusercontent.com/assets/7840213/26056667/3a5aa88c-396e-11e7-8459-50025f4884ac.png)

This seems to be because the extra feature info is being added *after* the closing `</table>` tag, instead of before it. See this from Chrome dev tools:

![2017-05-15 13_05_52-index html](https://cloud.githubusercontent.com/assets/7840213/26056968/62135094-396f-11e7-9e7e-a66b13269cf7.png)

I don't quite yet understand why this is happening. Can you have a look into it?

OlaKov commented:
Thanks for noticing, I was testing with photos in the popups and they showed up one after another. I got it fixed, and I simplified the code a bit by only checking if (doPopup) once. 

What would be the best way to handle clusters, right now it's not clear where the info for one feature ends and the other begins. We can do something like this:

"4 feature cluster:
Feature 1: 
Info
info
Feature 2:
info 
info" 

I don't know if "feature" and "cluster" are terms that are too technical?

![clusters](https://cloud.githubusercontent.com/assets/28122345/26068154/cb6acf6a-3959-11e7-8875-95b7ae1365dc.jpg)

tomchadwin commented:
I'd do it via alternate background shading alternate feature records ("zebra/candy-striping"). We cam so that in the CSS, if we change the HTML a little. I suggest we wrap the popup contents in a `<ul>`, and put each feature info into its own separate table in `<li>`s.

OlaKov commented:
Ah yes, a much better idea. I can make two `.li` classes (white and light gray) and alternate them in javascript. Should we limit the height of the popup to 400-500px, in case there are hundreds of fields?

tomchadwin commented:
No need for classes and JS. Just get the HTML right, and I'll sort the CSS.

OlaKov commented:
Something like this:

![clusters_ul](https://cloud.githubusercontent.com/assets/28122345/26089056/082e38de-39b8-11e7-86e1-8b89a35b3e71.jpg)
![clusters_ul_ul](https://cloud.githubusercontent.com/assets/28122345/26089067/23f2283c-39b8-11e7-8319-b504715c6f4e.jpg)

tomchadwin commented:
Precisely like that.

tomchadwin commented:
Try adding this to `resources/qgis2web.css`:

```css
#popup-content>ul>li:nth-child(odd) {
    background-color: #eee;
}
```

OlaKov commented:
That looks great:
![clusters_ul2](https://cloud.githubusercontent.com/assets/28122345/26117576/402e4ce4-3a24-11e7-980c-11920378c4cb.jpg)

I changed it to even, so that the first record is white and removed `<li>` decorations and `padding-left` does this seem right?
```
#popup-content>ul>li:nth-child(even) {
    background-color: #eee;
}
#popup-content ul {
    list-style-type: none;
    padding-left: 0;
}
```

tomchadwin commented:
Looking really good, and nearly there. Two issues:

1. The tables aren't quite working: the columns in each attribute row within each feature are not lining up - can't tell why until I can get to a laptop (on phone at the mo)

2. Can you add a little top/bottom margin to the `<li>`s or `<table>`s to separate the features? 0.25em or so should do it

OlaKov commented:
Ok sounds good, I'm gonna commit so we're on the same code.

tomchadwin commented:
Great. Can't guarantee when I'll get to it, I'm afraid. Big install starts at work tomorrow. But I'll do my best.



On Tue, May 16, 2017 at 6:58 PM +0100, "Ola Kovalchuk" <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok sounds good, I'm gonna commit so we're on the same code.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/tomchadwin/qgis2web/pull/492#issuecomment-301863513>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHeh1RJChAB1SZVTWDBraK560PoihhvCks5r6eOrgaJpZM4NaP_B>.



Tom Chadwin, ICT Manager
Telephone: 01434 611530 Mob:
Web: www.northumberlandnationalpark.org.uk<http://www.northumberlandnationalpark.org.uk/>

IMPORTANT NOTICE - Disclaimer - This communication is from Northumberland National Park Authority (NNPA).The Authority’s head office and principal place of business is Eastburn, South Park, Hexham, Northumberland, NE46 1BS, United Kingdom. If you are not the intended recipient(s) please note that any form of disclosure, distribution, copying or use of this communication or the information in it or in any attachments is strictly prohibited and may be unlawful. If you have received this communication in error, please delete the email and destroy any copies of it. Any views or opinions presented are solely those of the author and do not necessarily represent those of NNPA.Contractors or potential contractors are reminded that a formal Order or Contract is needed for NNPA to be bound by any offer or acceptance of terms for the supply of goods or services Although this email and any attachments are believed to be free of any virus or other defects which might affect any computer or IT system into which they are received, no responsibility is accepted by the NNPA for any loss or damage arising in any way from the receipt or use thereof. Computer systems of this Authority may be monitored and communications carried out on them recorded, to secure the effective operation of the system and for other lawful purpose.

OlaKov commented:
Yes, the columns are correct now, sorry about that
![clusters_ul4](https://cloud.githubusercontent.com/assets/28122345/26122294/3b9e03da-3a34-11e7-97a2-5b4a1f454a21.jpg)

In the browser it's displaying correctly, but in the QGIS preview the horizontal scroll appears.

![clusters_ul5](https://cloud.githubusercontent.com/assets/28122345/26122389/8b8a6000-3a34-11e7-9627-c7180f474135.jpg)

This is probably a css issue. I can look into it and also fix the tests, and you can merge tomorrow when you have time?

tomchadwin commented on olStyleScripts.py:
```diff
@@ -211,21 +211,21 @@ def exportStyles(layers, folder, clustered):
-        var labelText = String(%(label)s);
+    var labelText = ""
+    if (size >= 2) {
+        labelText = size.toString()
```
Doesn't this affect features labelled in QGIS?

tomchadwin commented on olStyleScripts.py:
```diff
@@ -211,21 +211,21 @@ def exportStyles(layers, folder, clustered):
-              offsetY: 3,
+              textAlign: "center",
+              offsetX: 0,
+              offsetY: 0,
```
As above, doesn't this affect labels from QGIS?

tomchadwin commented:
Confirmed re: labels, I'm afraid. If we can get your new code to implement *only on clustered layers* then that should bring non-clustered layer labels back. I'm not very familiar with the OL cluster code, though. Can you determine whether a layer is clustered within the layer style JS?

OlaKov commented:
Ah yes, that's right. The way I check for if the layer is clustered is:
clusteredFeatures = feature.get("features");
this is either undefined (for non-clustered feature) or contains an array with features that are clustered. I can do a similar check in the style js...

OlaKov commented:
I checked for clustering as described above and it seems to work:
![clusters_ul6](https://cloud.githubusercontent.com/assets/28122345/26126813/4690e2de-3a44-11e7-8d0e-fb90e4083692.jpg)

![clusters_ul7](https://cloud.githubusercontent.com/assets/28122345/26126819/4d22fc54-3a44-11e7-9207-fc45c4243b5c.jpg)

It's not the most elegant solution, it would be better to have python generate different style js files based on if the layer is clustered or not. This is possible, but I'm not very familiar with the python files yet. For now this works.. I'm just going to look at it a bit more carefully before commiting.

tomchadwin commented:
> In the browser it's displaying correctly, but in the QGIS preview the horizontal scroll appears.

Don't worry too much about that. It could be simply a bug in the version of WebKit in Qt. There are other minor display glitches in the preview window, but I've documented them. The export's what's important.

> it would be better to have python generate different style js files based on if the layer is clustered or not

You're right, of course. I hadn't thought of that. There are no doubt other elements of the OL popup code to which the same point applies. I would recommend sticking with this approach for now, and looking to simplify the output as a future development.

tomchadwin commented:
OK, tested, and all looking good here. Only odd thing: I have nothing at all now appearing in my preview. Export is fine. Are you seeing that?

OlaKov commented:
Hmm that's strange, my preview and exports are fine. This probably means that I have an un-closed bracket or a missing semicolon somewhere. I'll take a look..

tomchadwin commented:
No, I don't think so. I've got no console errors. Very odd indeed.

OlaKov commented:
The preview doesn't show up, no matter if the layer is clustered or not? Do you think the issue is with the olStyleScripts.py? This is what I'm seeing (QGIS 2.18):

![clusters_ul8](https://cloud.githubusercontent.com/assets/28122345/26166981/b515484a-3af2-11e7-90a2-2d9ada007f68.jpg)

tomchadwin commented:
I don't know what the issue is. To try to recreate, can you remove your "Field Sites" layer, and just leave "airports"? Then try that clustered and unclustered. That will match what my setup is.

OlaKov commented:
I deleted the files in the qgis2web plugin folder and replaced them with the files from this branch. And I created a new project with just the "airports" shape file (airports.shp) and it still works for me. Would someone else be willing to test this for us? Do you other data that you can test?

tomchadwin commented:
I'll do the same from your branch (just to check you reloaded the plugin before retesting)?

tomchadwin commented:
It's most likely some remaining difference in my local copy. I'll test. I'm just in the middle of something on the Leaflet export side of things at the mo...

tomchadwin commented:
OK, I must have been tired when I got the no preview issue. Just tested now and all looks good. Are you happy with the state it's in now (other than the tests)?

Looking great!

OlaKov commented:
I think this is good for now.. I tested with lakes (polygons) and piplines (lines) and it seems to work fine in terms of labels and popups:
![clusters_ul9](https://cloud.githubusercontent.com/assets/28122345/26214748/bf38ae5c-3bba-11e7-864b-db82e3fbb604.jpg)

Next, I would like to work on limiting the size of the circle, so we don't end up with a 500px circle for a 500 point cluster. It would be a change to the  olStyleScripts.py, but I'd need to look at it more carefully as there will be few functions affected.

tomchadwin commented:
Are you able to fix the tests before I merge?

OlaKov commented:
Yes, I will do this over the next hour or so.

tomchadwin commented:
Thanks so much. I'd like to consider a release soon. Do you think we should wait until you've had a chance to look at the oversize cluster issue?

I am uncertain whether to release the next version as experimental in the QGIS plugins repo because of the Leaflet styling work I've been doing, which has resulted in pretty wide-ranging changes on the Leaflet side of things. I can't quite decide.

OlaKov commented:
Sorry for so many commits, I can never get these tests quite right. I'm trying to merge some of the commits, but it's not working well. I'm doing the git rebase -i HEAD~4, which is suppose to open my text editor, but nothing is opening... 

Regarding a new version: I'm not sure... It will probably take me a week or a bit more to figure out the oversize cluster issue. I have some other responsibilities that are not as fun as this, but have to get done.

OlaKov commented:
No problem, always glad to help!
